File: fullstackProject/services/pricing-service/internal/domain/pricing.go
// Package domain — правила ценообразования Тарификатора: чистые
// функции без знания о транспортах и хранилищах.
package domain

import (
	"errors"
	"math"
	"time"
)

// Factor — слагаемое раскладки цены. Отсутствие фактора в ответе —
// множитель 1.0: раскладка показывает только то, что сработало.
type Factor struct {
	Code    string // машинное имя: evening, weekend, demand_full …
	Label   string // человекочитаемо: «вечерний прайм +20%»
	Percent int    // вклад фактора, %: −20 … +25
}

// Quote — расчёт цены одного места сеанса.
type Quote struct {
	BasePriceRub int      // базовая цена фильма из афиши
	PriceRub     int      // после факторов, кратно 10 ₽
	Factors      []Factor // что сработало, в порядке вычисления
	Occupied     int      // занятые места проекции спроса (после clamp)
	Capacity     int      // ёмкость зала
}

// QuoteRecord — проведённый расчёт для витрины (история квотов).
type QuoteRecord struct {
	SessionID    string
	BasePriceRub int
	PriceRub     int
	Factors      string // «evening+20%; weekend+10%» — компактная раскладка
	CreatedAt    time.Time
}

// DemandRecord — строка проекции спроса для витрины.
type DemandRecord struct {
	SessionID string
	Occupied  int
	UpdatedAt time.Time
}

var (
	// ErrInvalidQuote — запрос квота не проходит валидацию
	// (база ≤ 0, ёмкость ≤ 0, отрицательный спрос).
	ErrInvalidQuote = errors.New("некорректный запрос цены")
	// ErrInvalidDemandEvent — событие спроса без ключей или с нулевым
	// числом мест: обрабатывать нечего, ядовитое.
	ErrInvalidDemandEvent = errors.New("некорректное событие спроса")
)

// Пороги факторов. Часы — локальное время кинотеатра (TZ окружения:
// стенд задаёт Europe/Moscow в compose; зеркало на вебе считает
// в локальном времени браузера — на одном стенде это одинаковые часы).
const (
	// время суток: утро до 12 — скидка, вечер 17–23 — надбавка
	morningHourEnd = 12
	eveningHourBeg = 17
	nightHourBeg   = 23
	// спрос: доля занятых мест зала
	demandLowBelow  = 0.2 // меньше 20% — зал пустоват
	demandHighAbove = 0.5 // больше половины — спрос высокий
	demandFullAbove = 0.8 // больше 80% — аншлаг
)

// факторы времени суток: код, подпись, процент
func timeFactor(at time.Time) (Factor, bool) {
	h := at.Hour()
	switch {
	case h < morningHourEnd:
		return Factor{"morning", "утренний сеанс −20%", -20}, true
	case h >= eveningHourBeg && h < nightHourBeg:
		return Factor{"evening", "вечерний прайм +20%", +20}, true
	default:
		return Factor{}, false // день и поздняя ночь — базовый тариф
	}
}

// weekdayFactor — выходной (суббота, воскресенье) дорожает
func weekdayFactor(at time.Time) (Factor, bool) {
	wd := at.Weekday()
	if wd == time.Saturday || wd == time.Sunday {
		return Factor{"weekend", "выходной +10%", +10}, true
	}
	return Factor{}, false
}

// demandFactor —tier по доле занятых мест: аншлаг сильнее высокого
// спроса, пустой зал дешевле; середина (20–50%) — без фактора.
func demandFactor(occupied, capacity int) (Factor, bool) {
	share := float64(occupied) / float64(capacity)
	switch {
	case share >= demandFullAbove:
		return Factor{"demand_full", "аншлаг +25%", +25}, true
	case share >= demandHighAbove:
		return Factor{"demand_high", "спрос высокий +10%", +10}, true
	case share < demandLowBelow:
		return Factor{"demand_low", "зал почти пуст −10%", -10}, true
	default:
		return Factor{}, false
	}
}

// ComputeQuote — цена места: база × факторы, округление кратно 10 ₽.
// Чистая функция от входов: одинаковым входам — одинаковая цена,
// зеркало на вебе (shared/lib/pricing) считает так же.
func ComputeQuote(sessionAt time.Time, basePriceRub, occupied, capacity int) (Quote, error) {
	if basePriceRub <= 0 || capacity <= 0 {
		return Quote{}, ErrInvalidQuote
	}
	// проекция могла дрейфовать (потерянное событие) — зажимаем
	// в границы зала: доля спроса не выйдет за [0, 1]
	if occupied < 0 {
		occupied = 0
	}
	if occupied > capacity {
		occupied = capacity
	}

	local := sessionAt.In(time.Local)
	quote := Quote{
		BasePriceRub: basePriceRub,
		Occupied:     occupied,
		Capacity:     capacity,
		Factors:      []Factor{},
	}
	multiplier := 1.0
	apply := func(f Factor, ok bool) {
		if !ok {
			return
		}
		quote.Factors = append(quote.Factors, f)
		multiplier *= 1 + float64(f.Percent)/100
	}
	apply(timeFactor(local))
	apply(weekdayFactor(local))
	apply(demandFactor(occupied, capacity))
	quote.PriceRub = round10(float64(basePriceRub) * multiplier)
	return quote, nil
}

// round10 — цена места кратна десяти: монеты из афиши не возвращаются.
func round10(v float64) int {
	return int(math.Round(v/10) * 10)
}

File: fullstackProject/apps/web/src/shared/lib/pricing.ts
/**
 * Зеркало правил «Тарификатора» (services/pricing-service, Go).
 * Демо-движок считает цену теми же векторами, что и gRPC-сервис:
 * таблица проверки в pricing.spec.ts сверена с доменными тестами Go.
 * Часы сеанса — локальное время браузера: на стенде оно совпадает
 * с TZ сервиса (Europe/Moscow), на Pages демо живёт само в себе.
 */

/** Фактор раскладки цены: «вечерний прайм +20%» */
export interface PriceFactor {
  /** машинное имя: evening, weekend, demand_full … */
  code: 'morning' | 'evening' | 'weekend' | 'demand_low' | 'demand_high' | 'demand_full';
  /** человекочитаемая подпись — идентична Go-домену */
  label: string;
  /** вклад фактора, %: −20 … +25 */
  percent: number;
}

/** Расчёт цены места — как QuoteResponse Тарификатора */
export interface SessionQuote {
  sessionId: string;
  basePriceRub: number;
  priceRub: number;
  factors: PriceFactor[];
  occupied: number;
  capacity: number;
}

// пороги — как в internal/domain/pricing.go сервиса
const MORNING_HOUR_END = 12;
const EVENING_HOUR_BEG = 17;
const NIGHT_HOUR_BEG = 23;
const DEMAND_LOW_BELOW = 0.2;
const DEMAND_HIGH_ABOVE = 0.5;
const DEMAND_FULL_ABOVE = 0.8;

function timeFactor(at: Date): PriceFactor | null {
  const h = at.getHours();
  if (h < MORNING_HOUR_END) {
    return { code: 'morning', label: 'утренний сеанс −20%', percent: -20 };
  }
  if (h >= EVENING_HOUR_BEG && h < NIGHT_HOUR_BEG) {
    return { code: 'evening', label: 'вечерний прайм +20%', percent: 20 };
  }
  return null; // день и поздняя ночь — базовый тариф
}

function weekdayFactor(at: Date): PriceFactor | null {
  const wd = at.getDay(); // 0 — воскресенье, 6 — суббота
  if (wd === 0 || wd === 6) {
    return { code: 'weekend', label: 'выходной +10%', percent: 10 };
  }
  return null;
}

function demandFactor(occupied: number, capacity: number): PriceFactor | null {
  const share = occupied / capacity;
  if (share >= DEMAND_FULL_ABOVE) {
    return { code: 'demand_full', label: 'аншлаг +25%', percent: 25 };
  }
  if (share >= DEMAND_HIGH_ABOVE) {
    return { code: 'demand_high', label: 'спрос высокий +10%', percent: 10 };
  }
  if (share < DEMAND_LOW_BELOW) {
    return { code: 'demand_low', label: 'зал почти пуст −10%', percent: -10 };
  }
  return null;
}

/** Цена кратно десяти — как round10 в Go (half away from zero) */
function round10(v: number): number {
  return Math.round(v / 10) * 10;
}

/**
 * Цена места сеанса: база афиши × факторы времени суток, выходного
 * и заполненности зала. Чистая функция — одинаковые входы дают
 * одинаковую цену в демо и в live.
 */
export function quoteSession(input: {
  sessionId: string;
  /** момент сеанса, ISO */
  startsAt: string;
  basePriceRub: number;
  /** занятые места (в демо — карта занятости движка) */
  occupied: number;
  capacity: number;
}): SessionQuote {
  const occupied = Math.min(Math.max(input.occupied, 0), input.capacity);
  const at = new Date(input.startsAt);
  const factors: PriceFactor[] = [];
  let multiplier = 1;
  for (const factor of [timeFactor(at), weekdayFactor(at), demandFactor(occupied, input.capacity)]) {
    if (factor) {
      factors.push(factor);
      multiplier *= 1 + factor.percent / 100;
    }
  }
  return {
    sessionId: input.sessionId,
    basePriceRub: input.basePriceRub,
    priceRub: round10(input.basePriceRub * multiplier),
    factors,
    occupied,
    capacity: input.capacity,
  };
}
